File: server/services/dailySummaryService.ts
import { db } from "../db.js";
import { sales, borrowings, expenses } from "../../shared/schema.js";
import { gte, lte, and, eq } from "drizzle-orm";
import { notificationService } from "./notificationService.js";

interface DailySummary {
  date: Date;
  totalSales: number;
  totalExpenses: number;
  netProfit: number;
  newBorrowings: number;
  collectionsMade: number;
  overdueCount: number;
  summary: string;
}

class DailySummaryService {
  /**
   * Generate daily summary (with optional tenant filtering)
   */
  async generateDailySummary(date: Date = new Date(), mobileNo?: string): Promise<DailySummary> {
    const startOfDay = new Date(date);
    startOfDay.setHours(0, 0, 0, 0);
    const endOfDay = new Date(date);
    endOfDay.setHours(23, 59, 59, 999);

    try {
      // Get daily sales (filtered by mobileNo if provided)
      const dailySales = await db.query.sales.findMany({
        where: and(
          gte(sales.date, startOfDay),
          lte(sales.date, endOfDay),
          mobileNo ? eq(sales.mobileNo, mobileNo) : undefined
        ),
      });

      const totalSales = dailySales.reduce(
        (sum: number, s: any) => sum + parseFloat(s.amount.toString()),
        0
      );

      // Get daily expenses (filtered by mobileNo if provided)
      const dailyExpenses = await db.query.expenses.findMany({
        where: and(
          gte(sales.date, startOfDay),
          lte(sales.date, endOfDay),
          mobileNo ? eq(sales.mobileNo, mobileNo) : undefined
        ),
      });

      const totalExpenses = dailyExpenses.reduce(
        (sum: number, e: any) => sum + parseFloat(e.amount.toString()),
        0
      );

      // Get new borrowings (filtered by mobileNo if provided)
      const newBorrowings = await db.query.borrowings.findMany({
        where: and(
          gte(borrowings.date, startOfDay),
          lte(borrowings.date, endOfDay),
          eq(borrowings.status, "PENDING"),
          mobileNo ? eq(borrowings.mobileNo, mobileNo) : undefined
        ),
      });

      const totalNewBorrowings = newBorrowings.length;

      // Get paid borrowings (collections) (filtered by mobileNo if provided)
      const paidBorrowings = await db.query.borrowings.findMany({
        where: and(
          gte(borrowings.date, startOfDay),
          lte(borrowings.date, endOfDay),
          eq(borrowings.status, "PAID"),
          mobileNo ? eq(borrowings.mobileNo, mobileNo) : undefined
        ),
      });

      const totalCollections = paidBorrowings.reduce(
        (sum: number, b: any) => sum + parseFloat(b.amount.toString()),
        0
      );

      // Get overdue borrowings (filtered by mobileNo if provided)
      const now = new Date();
      const overdueBorrowings = await db.query.borrowings.findMany({
        where: and(
          eq(borrowings.status, "OVERDUE"),
          lte(borrowings.dueDate, now),
          mobileNo ? eq(borrowings.mobileNo, mobileNo) : undefined
        )
      });

      const netProfit = totalSales - totalExpenses;

      const summary = this.generateSummaryText(
        totalSales,
        totalExpenses,
        netProfit,
        totalNewBorrowings,
        totalCollections,
        overdueBorrowings.length,
        date
      );

      return {
        date,
        totalSales,
        totalExpenses,
        netProfit,
        newBorrowings: totalNewBorrowings,
        collectionsMade: totalCollections,
        overdueCount: overdueBorrowings.length,
        summary,
      };
    } catch (error) {
      console.error("Failed to generate daily summary:", error);
      throw error;
    }
  }

  /**
   * Generate summary text
   */
  private generateSummaryText(
    sales: number,
    expenses: number,
    profit: number,
    borrowings: number,
    collections: number,
    overdue: number,
    date: Date
  ): string {
    const dateStr = date.toLocaleDateString("en-US", {
      weekday: "long",
      year: "numeric",
      month: "long",
      day: "numeric",
    });

    return `
Daily Summary - ${dateStr}

📊 Sales & Revenue:
• Total Sales: ₹${sales.toFixed(2)}
• Total Expenses: ₹${expenses.toFixed(2)}
• Net Profit: ₹${profit.toFixed(2)}

💰 Credit Management:
• New Borrowings: ${borrowings}
• Collections Made: ₹${collections.toFixed(2)}
• Overdue Payments: ${overdue}

${profit >= 0 ? "✅ Profitable day!" : "⚠️ Loss recorded today."}

---
Generated: ${new Date().toLocaleTimeString()}
    `.trim();
  }

  /**
   * Send daily summary via SMS/WhatsApp
   */
  async sendDailySummary(
    phoneNumber: string,
    summary: DailySummary
  ): Promise<boolean> {
    try {
      const message = summary.summary;

      // Send via WhatsApp
      const result = await notificationService.sendWhatsAppMessage(
        phoneNumber,
        message
      );

      if (result.success) {
        // Log the notification
        await db.query.notificationsLog.findMany(); // Just to use db
        return true;
      }

      return false;
    } catch (error) {
      console.error("Failed to send daily summary:", error);
      return false;
    }
  }

  /**
   * Schedule daily summary for a specific time
   * This should be called by a cron job or scheduler
   */
  async scheduleDailySummary(phoneNumber: string, hour: number = 20, minute: number = 0) {
    const now = new Date();
    const scheduled = new Date();
    scheduled.setHours(hour, minute, 0, 0);

    // If time has passed today, schedule for tomorrow
    if (scheduled <= now) {
      scheduled.setDate(scheduled.getDate() + 1);
    }

    const timeUntilSummary = scheduled.getTime() - now.getTime();

    // Schedule the summary
    setTimeout(async () => {
      try {
        const summary = await this.generateDailySummary();
        await this.sendDailySummary(phoneNumber, summary);

        // Reschedule for next day
        this.scheduleDailySummary(phoneNumber, hour, minute);
      } catch (error) {
        console.error("Error sending scheduled summary:", error);
      }
    }, timeUntilSummary);

    return {
      success: true,
      scheduledFor: scheduled,
      timeUntilSummary: timeUntilSummary / (1000 * 60), // minutes
    };
  }

  /**
   * Get weekly summary (7-day report) (with optional tenant filtering)
   */
  async getWeeklySummary(mobileNo?: string): Promise<any> {
    const summaries: any[] = [];
    const now = new Date();

    for (let i = 6; i >= 0; i--) {
      const date = new Date(now);
      date.setDate(date.getDate() - i);
      const summary = await this.generateDailySummary(date, mobileNo);
      summaries.push(summary);
    }

    const totalWeeklySales = summaries.reduce((sum: number, s: any) => sum + s.totalSales, 0);
    const totalWeeklyExpenses = summaries.reduce((sum: number, s: any) => sum + s.totalExpenses, 0);
    const totalWeeklyProfit = summaries.reduce((sum: number, s: any) => sum + s.netProfit, 0);
    const totalWeeklyCollections = summaries.reduce((sum: number, s: any) => sum + s.collectionsMade, 0);

    return {
      week: `${(summaries[0] as any).date.toLocaleDateString()} - ${(summaries[6] as any).date.toLocaleDateString()}`,
      dailySummaries: summaries,
      totalSales: totalWeeklySales,
      totalExpenses: totalWeeklyExpenses,
      totalProfit: totalWeeklyProfit,
      totalCollections: totalWeeklyCollections,
      averageDailySales: totalWeeklySales / 7,
      bestDay: summaries.reduce((max: any, s: any) => (s.totalSales > max.totalSales ? s : max)),
      worstDay: summaries.reduce((min: any, s: any) => (s.totalSales < min.totalSales ? s : min)),
    };
  }
}

export const dailySummaryService = new DailySummaryService();

File: backend/src/main/java/com/apnashop/service/DailySummaryService.java
package com.apnashop.service;

import com.apnashop.dto.DailySummaryDto;
import com.apnashop.dto.WeeklySummaryDto;
import com.apnashop.entity.Borrowing;
import com.apnashop.entity.BorrowingStatus;
import com.apnashop.entity.Sale;
import com.apnashop.repository.BorrowingRepository;
import com.apnashop.repository.SaleRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.time.format.TextStyle;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.ScheduledThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

@Service
@RequiredArgsConstructor
public class DailySummaryService {

    private final SaleRepository saleRepository;
    private final BorrowingRepository borrowingRepository;
    private final ScheduledThreadPoolExecutor scheduler = new ScheduledThreadPoolExecutor(1);
    private final Map<String, ScheduledFuture<?>> scheduledJobs = new ConcurrentHashMap<>();

    public DailySummaryDto generateDailySummary(LocalDate date, String mobileNo) {
        Instant startOfDay = date.atStartOfDay(ZoneOffset.UTC).toInstant();
        Instant endOfDay = date.plusDays(1).atStartOfDay(ZoneOffset.UTC).toInstant().minusMillis(1);

        List<Sale> dailySales = saleRepository.findByMobileNoAndDateBetween(mobileNo, startOfDay, endOfDay);
        double totalSales = dailySales.stream()
                .mapToDouble(s -> s.getAmount().doubleValue())
                .sum();

        List<Borrowing> newBorrowings = borrowingRepository.findByMobileNoAndDateBetweenAndStatus(
                mobileNo, startOfDay, endOfDay, BorrowingStatus.PENDING);
        long newBorrowingsCount = newBorrowings.size();

        List<Borrowing> paidBorrowings = borrowingRepository.findByMobileNoAndDateBetweenAndStatus(
                mobileNo, startOfDay, endOfDay, BorrowingStatus.PAID);
        double collectionsMade = paidBorrowings.stream()
                .mapToDouble(b -> b.getAmount().doubleValue())
                .sum();

        List<Borrowing> overdueBorrowings = borrowingRepository.findByMobileNoAndStatusAndDueDateLessThanEqual(
                mobileNo, BorrowingStatus.OVERDUE, Instant.now());
        long overdueCount = overdueBorrowings.size();

        double totalExpenses = 0;
        double netProfit = totalSales - totalExpenses;
        String summary = buildSummaryText(date, totalSales, totalExpenses, netProfit,
                newBorrowingsCount, collectionsMade, overdueCount);

        return DailySummaryDto.builder()
                .date(startOfDay)
                .totalSales(totalSales)
                .totalExpenses(totalExpenses)
                .netProfit(netProfit)
                .newBorrowings(newBorrowingsCount)
                .collectionsMade(collectionsMade)
                .overdueCount(overdueCount)
                .summary(summary)
                .build();
    }

    public WeeklySummaryDto getWeeklySummary(String mobileNo) {
        List<DailySummaryDto> summaries = new ArrayList<>();
        LocalDate today = LocalDate.now(ZoneOffset.UTC);

        for (int i = 6; i >= 0; i--) {
            summaries.add(generateDailySummary(today.minusDays(i), mobileNo));
        }

        double totalSales = summaries.stream().mapToDouble(DailySummaryDto::getTotalSales).sum();
        double totalExpenses = summaries.stream().mapToDouble(DailySummaryDto::getTotalExpenses).sum();
        double totalProfit = summaries.stream().mapToDouble(DailySummaryDto::getNetProfit).sum();
        double totalCollections = summaries.stream().mapToDouble(DailySummaryDto::getCollectionsMade).sum();

        DailySummaryDto bestDay = summaries.stream()
                .max(Comparator.comparingDouble(DailySummaryDto::getTotalSales))
                .orElse(summaries.getFirst());
        DailySummaryDto worstDay = summaries.stream()
                .min(Comparator.comparingDouble(DailySummaryDto::getTotalSales))
                .orElse(summaries.getFirst());

        String week = formatDate(summaries.getFirst().getDate()) + " - "
                + formatDate(summaries.getLast().getDate());

        return WeeklySummaryDto.builder()
                .week(week)
                .dailySummaries(summaries)
                .totalSales(totalSales)
                .totalExpenses(totalExpenses)
                .totalProfit(totalProfit)
                .totalCollections(totalCollections)
                .averageDailySales(totalSales / 7.0)
                .bestDay(bestDay)
                .worstDay(worstDay)
                .build();
    }

    public boolean sendDailySummary(String phoneNumber, DailySummaryDto summary) {
        // WhatsApp/SMS integration not configured; log for now.
        return phoneNumber != null && !phoneNumber.isBlank() && summary.getSummary() != null;
    }

    public Map<String, Object> scheduleDailySummary(String phoneNumber, int hour, int minute) {
        Instant now = Instant.now();
        LocalDate today = LocalDate.now(ZoneOffset.UTC);
        Instant scheduledInstant = today.atTime(hour, minute).toInstant(ZoneOffset.UTC);
        if (!scheduledInstant.isAfter(now)) {
            scheduledInstant = today.plusDays(1).atTime(hour, minute).toInstant(ZoneOffset.UTC);
        }

        long delayMs = scheduledInstant.toEpochMilli() - now.toEpochMilli();
        String key = phoneNumber + ":" + hour + ":" + minute;

        scheduledJobs.computeIfAbsent(key, k -> scheduler.schedule(
                () -> sendDailySummary(phoneNumber, generateDailySummary(LocalDate.now(ZoneOffset.UTC), phoneNumber)),
                delayMs,
                TimeUnit.MILLISECONDS
        ));

        return Map.of(
                "success", true,
                "scheduledFor", scheduledInstant.toString(),
                "timeUntilSummary", delayMs / (1000.0 * 60.0)
        );
    }

    private String buildSummaryText(
            LocalDate date,
            double sales,
            double expenses,
            double profit,
            long borrowings,
            double collections,
            long overdue
    ) {
        String dateStr = date.getDayOfWeek().getDisplayName(TextStyle.FULL, Locale.US) + ", "
                + date.getMonth().getDisplayName(TextStyle.FULL, Locale.US) + " "
                + date.getDayOfMonth() + ", " + date.getYear();

        return """
                Daily Summary - %s
                
                📊 Sales & Revenue:
                • Total Sales: ₹%.2f
                • Total Expenses: ₹%.2f
                • Net Profit: ₹%.2f
                
                💰 Credit Management:
                • New Borrowings: %d
                • Collections Made: ₹%.2f
                • Overdue Payments: %d
                
                %s
                
                ---
                Generated: %s
                """.formatted(
                dateStr,
                sales,
                expenses,
                profit,
                borrowings,
                collections,
                overdue,
                profit >= 0 ? "✅ Profitable day!" : "⚠️ Loss recorded today.",
                Instant.now().toString()
        ).trim();
    }

    private String formatDate(Instant instant) {
        return DateTimeFormatter.ISO_LOCAL_DATE.format(instant.atZone(ZoneOffset.UTC));
    }
}
